urllib3/sessionmanager.py

```diff
@@ -0,0 +1,81 @@
+from .util.retry import Retry
+from .request import RequestMethods, Request
+from .util.context import Context
+
+
+class SessionManager(RequestMethods):
+    """
+    Allows arbitrary requests while maintaining session context across
+    those requests. Currently, that context consists of automatic
+    cookie storage and retrieval.
+
+    :param manager:
+        An appropriate :class:`PoolManager` or :class:`ProxyManager` object
+        to handle HTTP requests for the SessionManager
+
+    :param cookie_jar:
+        A predefined :class:`CookieJar` object to use in the session;
+        if not provided, a new one will be created.
+
+    :param headers:
+        A set of headers to include with all requests, unless other
+        headers are given explicitly.
+
+    """
+    def __init__(self, manager, cookie_jar=None, headers=None):
+        super(SessionManager, self).__init__(headers=headers)
+        self.manager = manager
+        self.context = Context(cookie_jar)
+
+    def set_cookie_policy(self, policy):
+        """
+        Set the encapsulated context's cookie policy
+        """
+        self.context.set_cookie_policy(policy)
+
+    def apply_context(self, request):
+        """
+        Take the given request object and mutate it to fit the current context
+        of the SessionManager
+        """
+        self.context.apply_to(request)
+
+    def extract_context(self, response, request):
+        """
+        Take the given response and the request that produced it, and store
+        any context it produced for later retrieval.
+        """
+        self.context.extract_from(response, request)
+
+    def urlopen(self, method, url, redirect=True, retries=None, **kw):
```

Comment:
All right; should we treat this note as the first step in moving redirect (and other) logic into `RequestMethods.request()`, or do you still want to think about doing that inside this PR?

My thought on that would be that this can serve as a notice that we're going to begin removing "smart" features like redirect from `urlopen`, and we can actually implement that stripping in the next major revision.